[Azure Maps] Weather: Swagger improvement for SDK generation [Hub Generated]

alextts627 opened the pull request:
This is a PR generated at [OpenAPI Hub](https://aka.ms/openapihub). You can view your [work branch via this link](https://portal.azure-devex-tools.com/branch/alankashiwa/azure-rest-api-specs/maps-geolocation-improvement-for-sdk...main/).


### Changelog
Add a changelog entry for this PR by answering the following questions:
  1. What's the purpose of the update?
      - [x] **To improve swagger for generated clients and address some issues from [this review in ApiView](https://apiview.dev/Assemblies/Review/9a6ea029a8ef4aafaa52fdbf12695b21)**
  2. When are you targeting to deploy the new service/feature to public regions? **Services already deployed.**
  3. When do you expect to publish the swagger? **ASAP**
  4. If updating an existing version, please select the specific language SDKs and CLIs that must be refreshed after the swagger is published. **No data-plane SDKs are onboarded yet, so no refresh is required.**

### Contribution checklist:
- [x] I commit to follow the [Breaking Change Policy](http://aka.ms/bcforapi) of "no breaking changes"
- [x] I have reviewed the [documentation](https://aka.ms/ameonboard) for the workflow.
- [x] [Validation tools](https://aka.ms/swaggertools) were run on swagger spec(s) and errors have all been fixed in this PR. [How to fix?](https://aka.ms/ci-fix)

If any further questions about AME onboarding or validation tools, please view the [FAQ](https://aka.ms/faqinprreview).

### Breaking Change Review Checklist 
If any of the following scenarios apply to the PR, request approval from the Breaking Change Review Board as defined in the [Breaking Change Policy](http://aka.ms/bcforapi). 
- [ ] Removing API(s) in a stable version
- [ ] Removing properties in a stable version
- [ ] Removing API version(s) in a stable version
- [x] Updating API in a stable or public preview version with Breaking Change Validation errors
- [ ] Updating API(s) in public preview over 1 year (refer to [Retirement of Previews](https://dev.azure.com/msazure/AzureWiki/_wiki/wikis/AzureWiki.wiki/37683/Retirement-of-Previews))

**Action**: to initiate an evaluation of the breaking change, create a new intake using the [template for breaking changes](https://aka.ms/Breakingchangetemplate). Addition details on the process and office hours are on the [Breaking change Wiki](https://dev.azure.com/msazure/AzureWiki/_wiki/wikis/AzureWiki.wiki/37684/Breaking-Changes).

AlexTsao committed:
Modified AirQuality

AlexTsao committed:
Modified AirDetails

alextts627 committed:
renaming variables for Current Condition

alextts627 committed:
1. Rename to model to WeatherValue

2. Renamed to IconCode definition

alextts627 committed:
Rename properties to description(s) for consistency

alextts627 committed:
Rename model to weatherValueRange

alextts627 committed:
Renamed to Forecast for better self explaining

alextts627 committed:
Better naming/self explaining for values

alextts627 committed:
Update name fpr dbz

alextts627 committed:
Add enum for UnitType, PrecipitationType, IconCode

alankashiwa commented on specification/maps/data-plane/Weather/preview/1.0/weather.json:
```diff
@@ -1282,10 +1282,10 @@
       "type": "object",
       "properties": {
-        "name": {
+        "description": {
```
You cannot just change the name directly otherwise the deserialization of API response may fail in some languages I would assume. Should use `x-ms-client-name` instead.

alankashiwa replied:
And there is a property referencing `AirAndPollen` as well.

alankashiwa replied:
There are other fields under "properties" section where `x-ms-client-name` should be used.

alextts627 replied:
yeah I have been trying to fix these check errors. Thanks for reminding :)

alankashiwa commented on specification/maps/data-plane/Weather/preview/1.0/weather.json:
```diff
@@ -1303,7 +1303,7 @@
           "description": "Value associated with the air quality or pollution category. These values range from 1 to 6. 1 implying good conditions, 6 implying hazardous conditions."
         },
-        "type": {
+        "air_quality_type": {
```
Should use x-ms-client-name instead. And `airQualityType` (will become `air_quality_type` in Python)

alankashiwa commented on specification/maps/data-plane/Weather/preview/1.0/weather.json:
```diff
@@ -502,7 +502,429 @@
+          },
+          {
+            "value": 11,
+            "description": "hectoPascals"
```
This unknown word fails [Spell check](https://github.com/Azure/azure-rest-api-specs/pull/16148/checks?check_run_id=3751822770).  There is an instruction on how to fix that.

alankashiwa replied:
[Prettier formatting](https://github.com/Azure/azure-rest-api-specs/pull/16148/checks?check_run_id=3751822760) for the Swagger should be rerun as well.

alextts627 committed:
Updated example files

alextts627 committed:
Fix CI errors

1. Add new word to avoid error checks failing
2. Updated with "x-ms-client-name" to fix CI errors

daniel-rocha commented:
@alextts627 Please run prettier locally prior to pushing. There are instructions on how to do it. 

Model validation should also be run locally: https://github.com/Azure/oav

alankashiwa commented on specification/maps/data-plane/Weather/preview/1.0/weather.json:
```diff
@@ -502,7 +502,430 @@
+          },
+          {
+            "value": 36,
+            "name": "Intermittent Clouds",
```
Duplicated with `"value": 4`.
Other duplicates: (1) 6 and 38, (2) 16 and 42 (3) 20 and 43 (4) 23 and 44 (5) 34 and 40.

TypeScript SDK build will fail because of the duplicated values.

alextts627 replied:
updated, thanks for checking the JS for me.

alankashiwa commented on specification/maps/data-plane/Weather/preview/1.0/weather.json:
```diff
@@ -512,12 +935,12 @@
-          "type": "integer",
-          "format": "int32",
+          "$ref": "#/definitions/UnitType",
           "description": "Numeric ID value associated with the type of unit being displayed. Can be used for unit translation. Please refer to [Weather Service Concepts](https://aka.ms/AzureMapsWeatherConcepts) for details."
```
This `description` can be deleted since it is also in the `UnitType` definition.

alankashiwa commented on specification/maps/data-plane/Weather/preview/1.0/weather.json:
```diff
@@ -565,13 +989,13 @@
-          "type": "integer",
-          "format": "int32",
+          "$ref": "#/definitions/UnitType",
           "description": "Numeric ID value associated with the type of unit being displayed. Can be used for unit translation. Please refer to [Weather Service Concepts](https://aka.ms/AzureMapsWeatherConcepts) for details."
```
This `description` can be deleted since it is also in the `UnitType` definition.

alankashiwa commented on specification/maps/data-plane/Weather/preview/1.0/weather.json:
```diff
@@ -593,19 +1017,52 @@
-          "type": "integer",
-          "format": "int32",
+          "$ref": "#/definitions/IconCode",
           "description": "Numeric value representing an image that displays the `iconPhrase`. Please refer to [Weather Service Concepts](https://aka.ms/AzureMapsWeatherConcepts) for details."
```
This `description` can be deleted since it is also in the `IconCode` definition.

alankashiwa commented on specification/maps/data-plane/Weather/preview/1.0/weather.json:
```diff
@@ -739,8 +1200,7 @@
-          "type": "integer",
-          "format": "int32",
+          "$ref": "#/definitions/IconCode",
           "description": "Numeric value representing an image that displays the `iconPhrase`. Please refer to [Weather Service Concepts](https://aka.ms/AzureMapsWeatherConcepts) for details."
```
This `description` can be deleted since it is also in the `IconCode` definition.

alankashiwa commented on specification/maps/data-plane/Weather/preview/1.0/weather.json:
```diff
@@ -1335,23 +1799,27 @@
             "$ref": "#/definitions/AirAndPollen"
           },
-          "description": "airAndPollen"
+          "description": "air quality"
```
Should at least capitalize the description to be consistent with sibling properties.

alankashiwa commented on specification/maps/data-plane/Weather/preview/1.0/weather.json:
```diff
@@ -1905,8 +2383,9 @@
+          "x-ms-client-name": "isAscending",
           "type": "boolean",
-          "description": "Describes the direction of the `value` and `categoryValue`. For example, when `ascending`=True, the poorest index value is 0 and the best index value is 10. When `ascending`=True, the poorest index value is 10 and the best index value is 0."
+          "description": "Describes the direction of the `value` and `categoryValue`. For example, when `is_ascending`=True, the poorest index value is 0 and the best index value is 10. When `is_ascending`=True, the poorest index value is 10 and the best index value is 0."
```
[Rest docs](https://docs.microsoft.com/en-us/rest/api/maps/weather/get-daily-indices#dailyindex) would become inconsistent here and in other descriptions modified...

alextts627 replied:
Should we just change the rest docs separately.

alankashiwa replied:
Rest docs also use the `description` here so you've already changed them.
I think the current status is as follow:

- API response: `ascending` (in production, cannot be changed)
- Rest docs: `is_ascending` (inconsistent with API response)
- Generated client and docstrings: `is_ascending` 

Either we don't rename `ascending`, or we overwrite the docstrings in the convenience layer.
@ambientlight @daniel-rocha Any thoughts on this?

tjprescott replied:
You could also simply reword this so it doesn't reference the name. 
"When set to true, the poorest..."

alextts627 committed:
Update according to review feedbacks

1. Removed duplicate enum attributes
2. Updated descriptions

alextts627 committed:
Update IconCode

tjprescott commented on specification/maps/data-plane/Weather/preview/1.0/weather.json:
```diff
@@ -502,7 +502,400 @@
+          },
+          {
+            "value": 8,
+            "name": "Dreary (Overcast)",
```
I'll be interested to see how this is converted by Autorest.

tjprescott commented on specification/maps/data-plane/Weather/preview/1.0/weather.json:
```diff
@@ -593,20 +985,52 @@
+      "x-ms-enum": {
+        "name": "HazardIndex",
+        "modelAsString": true,
+        "values": [
+          {
+            "value": 0,
+            "description": "No hazard."
+          },
+          {
+            "value": 1,
+            "description": "Be informed, be aware."
+          },
+          {
+            "value": 2,
+            "description": "Pay attention, be prepared."
+          },
+          {
+            "value": 3,
+            "description": "Take action."
+          },
+          {
+            "value": 4,
+            "description": "Life threatening, emergency."
+          }
+        ]
```
This most likely need short names to map to enum labels. Example: `EMERGENCY = 4`

alextts627 committed:
Update x-ms-exmaple for better description

alextts627 committed:
Update Enum

daniel-rocha commented:
Temperature Summary is also defined non-optimally. It is generating three different classes (TemperatureSummary6Hours, 12Hours, 24Hours) with the same properties. All of TemperatureSummary's fields need to be a single type. 

Final desired result:

`
TemperatureSummaryItem (or similar name)  t = TemperatureSummary.getPast6Hours();
TemperatureSummaryItem (or similar name) t2 = TemperatureSummary.getPast12Hours(); 
`

Also past6Hours is not a particularly great field name, I'd alias it with pastSixHours, pastTwelveHours, etc.

tjprescott reviewed (changes requested):
Some lingering issues with the int-backed enums but otherwise looks solid.

daniel-rocha reviewed (changes requested):
Changes asked in comments.

daniel-rocha reviewed (changes requested):
All **response** objects should be readOnly = true. In the current swagger, most (if not all) objects returning from the calls have setters, which is not allowed by the guidelines. Objects returning from service calls need to be immutable.

**Request** objects obviously need to have setters.

alextts627 committed:
add name for enum

alextts627 committed:
Update according to feedback

1. Update all `query` to collectionFormat
2. Set `readOnly=true` for all response
3. Create PastHoursTemperature model for opimization

alextts627 commented:
> @alextts627 All "query" parameters in all methods are incorrect. They should be collectionFormat csv with numbers type double. You can just copy the definition from search:
> 
> ```
>     "Position": {
>       "name": "query",
>       "in": "query",
>       "description": "The applicable query specified as a comma separated string composed by latitude followed by longitude e.g. \"47.641268,-122.125679\".",
>       "required": true,
>       "x-ms-parameter-location": "method",
>       "type": "array",
>       "collectionFormat": "csv",
>       "minItems": 2,
>       "maxItems": 2,
>       "items": {
>         "type": "number",
>         "format": "double"
>       }
>     },
> `
> 
> In the codegen, parameter should be a list of doubles, not a string. This is a blocker for this PR. 
> 
> Also, please change the name from "query" to "coordinates" x-ms-client-name since it's no longer a query. 
> 
> WeatherAlongRoute is an exception and I will comment separately
> ```

I have changed all the `name=query` to collectionFormat with `x-ms-client-name=coordinates`

alextts627 committed:
Update name
